### PR TITLE
Store info in Cholesky type

### PR DESCRIPTION
**Pull request by fredrikekre**
Implements https://github.com/JuliaLang/LinearAlgebra.jl/issues/422

The first commit is JuliaLang/julia#21595 (bump on that, its pure code deletion), so when that is merged I will rebase.

Tests pass locally, but there are still things that needs to be changed, for instance

```julia
julia> A = [1. 2.; 2. 1.];

julia> det(A)
-3.0

julia> det(cholfact(A))
9.0
```

### CODE REVIEW

**Comment by fredrikekre** [comment]
I think I will deal with the `isposdef` part of https://github.com/JuliaLang/julia/issues/21559#issuecomment-297424747 in a separate PR (it requires more changes than I first thought). Is this news worthy?

File: base/linalg/cholesky.jl [hunk 1]
```diff
@@ -99,30 +113,33 @@ function _chol!(A::AbstractMatrix, ::Type{LowerTriangular})
             end
         end
      end
-    return LowerTriangular(A)
+    return LowerTriangular(A), convert(BlasInt, 0) # TODO: If we get here, do we know A is pos. def?
```
Comment (andreasnoack) [inline base/linalg/cholesky.jl]: I'd think so. What could be wrong?

**Review by andreasnoack — APPROVED** [review]

File: base/linalg/cholesky.jl (hunk 1 above)
Comment (fredrikekre) [inline base/linalg/cholesky.jl]: Okay thanks. Just wanted to make sure, will remove the comment.

Comment (fredrikekre) [inline base/linalg/cholesky.jl]: I was paranoid so confirmed with experiments:
```julia
for N in (10, 100, 1000), i in 1:1000
    A = rand(N, N); B = A'A
    for M in (A, B)
        C, info = invoke(Base.LinAlg._chol!, Tuple{AbstractMatrix, Type{LowerTriangular}}, copy(M), LowerTriangular)
        infoblas = cholfact(Hermitian(M, :L)).info
        info == infoblas == 0 || (info > 0 && infoblas > 0) || error()
    end
    for M in (A, B)
        C, info = invoke(Base.LinAlg._chol!, Tuple{AbstractMatrix, Type{UpperTriangular}}, copy(M), UpperTriangular)
        infoblas = cholfact(Hermitian(M, :U)).info
        info == infoblas == 0 || (info > 0 && infoblas > 0) || error()
    end
end
```

File: test/linalg/cholesky.jl [hunk 2]
```diff
@@ -272,3 +271,14 @@ end
 
 # Fail for non-BLAS element types
 @test_throws ArgumentError cholfact!(Hermitian(rand(Float16, 5,5)), Val{true})
+
+@testset "throw for non positive matrix" begin
+    for T in (Float32, Float64, Complex64, Complex128)
+        A = T[1 2; 2 1]; B = T[1, 1]
+        C = cholfact(A)
+        @show typeof(A), typeof(B), typeof(C.factors)
```
Comment (tkelman) [inline test/linalg/cholesky.jl]: debugging output left in?

Comment (fredrikekre) [inline test/linalg/cholesky.jl]: yes, sry

**Pull request merged**